File: 0536-38-2349-design-a-number-container-system/0536-38-2349-design-a-number-container-system.js
class NumberContainers {
    constructor() {
        this.indexToNumber = new Map();
        this.numberToIndices = new Map();
    }

    change(index, number) {
        if (this.indexToNumber.has(index)) {
            let oldNumber = this.indexToNumber.get(index);
            if (this.numberToIndices.has(oldNumber)) {
                let arr = this.numberToIndices.get(oldNumber);
                let idx = arr.indexOf(index);
                if (idx !== -1) arr.splice(idx, 1); 
                if (arr.length === 0) this.numberToIndices.delete(oldNumber);
            }
        }

        this.indexToNumber.set(index, number);
        if (!this.numberToIndices.has(number)) this.numberToIndices.set(number, []);
        
        let arr = this.numberToIndices.get(number);
        let pos = this.binaryInsert(arr, index);  
        arr.splice(pos, 0, index);
    }

    find(number) {
        return this.numberToIndices.has(number) && this.numberToIndices.get(number).length > 0
            ? this.numberToIndices.get(number)[0] // Smallest index
            : -1;
    }

    binaryInsert(arr, val) {
        let left = 0, right = arr.length;
        while (left < right) {
            let mid = Math.floor((left + right) / 2);
            if (arr[mid] < val) left = mid + 1;
            else right = mid;
        }
        return left;
    }
}

File: 0536-38-2349-design-a-number-container-system/0536-38-2349-design-a-number-container-system.py
from collections import defaultdict
import sortedcontainers

class NumberContainers:
    def __init__(self):
        self.index_to_number = {}
        self.number_to_indices = defaultdict(sortedcontainers.SortedSet)

    def change(self, index: int, number: int) -> None:
        if index in self.index_to_number:
            old_number = self.index_to_number[index]
            self.number_to_indices[old_number].discard(index)
            if not self.number_to_indices[old_number]:
                del self.number_to_indices[old_number]

        self.index_to_number[index] = number
        self.number_to_indices[number].add(index)

    def find(self, number: int) -> int:
        return next(iter(self.number_to_indices[number]), -1)
